File: Dataset/Advanced/Test/OOP_prepared/Java/Book_3_trace_2.java
package Book3;

class Author {
    private String name;
    private String email;

    public Author(String name, String email) {
        this.name = name;
        this.email = email;
    }

    public Author() {
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    @Override
    public String toString() {
        return "Author[" + "name=" + name + ",email=" + email + ']';
    }
}

public class Book {
    private String isbn;
    private String name;
    private Author author;
    private Double price;
    private int qty;

    public Book(String isbn, String name, Author author, Double price) {
        this.isbn = isbn;
        this.name = name;
        this.author = author;
        this.price = price;
    }

    public Book(String isbn, String name, Author author, Double price, int qty) {
        this.isbn = isbn;
        this.name = name;
        this.author = author;
        this.price = price;
        this.qty = qty;
    }

    public String getIsbn() {
        return isbn;
    }

    public void setIsbn(String isbn) {
        this.isbn = isbn;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Author getAuthor() {
        return author;
    }

    public void setAuthor(Author author) {
        this.author = author;
    }

    public Double getPrice() {
        return price;
    }

    public void setPrice(Double price) {
        this.price = price;
    }

    public int getQty() {
        return qty;
    }

    public void setQty(int qty) {
        this.qty = qty;
    }
    
    public String getAuthorName(){
        return this.author.getName();
    }

    @Override
    public String toString() {
        return "Book[" + "isbn=" + isbn + ",name=" + name + "," + author + ",price=" + price + ",qty=" + qty + ']';
    } 
}

class TestBook {
    public static void main(String[] args) {
        Author a2 = new Author("Jane Smith", "jane@example.com");
        Book b2 = new Book("22222", "Advanced Java", a2, 25.0, 30);
        System.out.println(b2.getAuthorName());
    }
}

File: Dataset/Advanced/Test/OOP_prepared/Python/Book_3_trace_2.py
class Author:
    def __init__(self, name=None, email=None):
        self.name = name
        self.email = email

    def get_name(self):
        return self.name

    def set_name(self, name):
        self.name = name

    def get_email(self):
        return self.email

    def set_email(self, email):
        self.email = email

    def __str__(self):
        return f"Author[name={self.name},email={self.email}]"


class Book:
    def __init__(self, isbn, name, author, price, qty=0):
        self.isbn = isbn
        self.name = name
        self.author = author
        self.price = price
        self.qty = qty

    def get_isbn(self):
        return self.isbn

    def set_isbn(self, isbn):
        self.isbn = isbn

    def get_name(self):
        return self.name

    def set_name(self, name):
        self.name = name

    def get_author(self):
        return self.author

    def set_author(self, author):
        self.author = author

    def get_price(self):
        return self.price

    def set_price(self, price):
        self.price = price

    def get_qty(self):
        return self.qty

    def set_qty(self, qty):
        self.qty = qty

    def get_author_name(self):
        return self.author.get_name()

    def __str__(self):
        return f"Book[isbn={self.isbn},name={self.name},{self.author},price={self.price},qty={self.qty}]"


if __name__ == "__main__":
    author2 = Author("Jane Smith", "jane@example.com")
    book1 = Book("978-3-16-148410-0", "Python 101", author2, 29.99, 5)
    print(book1)
    print(book1.get_author_name())